### ISSUE TITLE
Mechanism for Observing Things, Properties, handling Thing Events and Discovery

### ISSUE THREAD

**Issue by zolkis** [w3c/wot-scripting-api#158]
This is a summary and clean-page continuation of #153.
The WoT Scripting API is a direct "low-level" mapping of the [WoT Thing Description](https://w3c.github.io/wot-thing-description/) spec and is primarily meant for implementations in  [constrained] web runtimes. However, we try to align it with the Web Platform, as browser implementations might be possible.

In order to 
- track [`ConsumedThing`](https://w3c.github.io/wot-scripting-api/#the-consumedthing-interface) Thing Description changes, 
- observe [`ThingProperty`](https://w3c.github.io/wot-scripting-api/#the-thingproperty-interface) changes, 
- handle [`ThingEvent`](https://w3c.github.io/wot-scripting-api/#the-thingevent-interface) notifications,
- handle [Thing discovery](https://w3c.github.io/wot-scripting-api/#the-discover-method), 
(pasting here the [first comment of #153 ](https://github.com/w3c/wot-scripting-api/issues/153#issue-376812090))

> Currently the WoT Scripting spec uses a minimal definition of `Observable`:
> 
> ```js
> interface Observable {
>   Subscription subscribe(EventHandler handler,
>                          optional ErrorHandler errorHandler,
>                          optional OnComplete onComplete);
> };
> interface Subscription {
>   void unsubscribe();
>   readonly attribute boolean closed;
> };
> callback EventHandler = void (any value);
> callback ErrorHandler = void (Error error);
> callback OnComplete = void ();
> ```
> 
> It is used by the `wot.discover()` method and it is implemented by [`ConsumedThing`](https://w3c.github.io/wot-scripting-api/#the-consumedthing-interface) for observing TD changes, [`ThingProperty`](https://w3c.github.io/wot-scripting-api/#the-thingproperty-interface) and [`ThingEvent`](https://w3c.github.io/wot-scripting-api/#the-thingevent-interface) for observing value changes.
> 
> However, verbal feedback on TPAC 2018 discouraged the use of Observables.
> 
> Instead, we should look into applying the [TAG design document](https://w3ctag.github.io/design-principles/#event-design) that encourages using [Event](https://dom.spec.whatwg.org/#event), [EventTarget](https://dom.spec.whatwg.org/#interface-eventtarget), [AbortController](https://dom.spec.whatwg.org/#interface-abortcontroller), and [AbortSignal](https://dom.spec.whatwg.org/#abortsignal).
> 
> There are several options forward (expanded in next comments).

Since then (also due to discussions on [Web NFC](https://github.com/w3c/web-nfc/issues/152)), we discarded the idea of using `AbortController` and aim using Promises for one-off and a simple solution (to be agreed upon in this issue) for multiple-off asynchronous notifications. At the moment we have a few alternatives for that "simple" solution, described next.

1. A simple new Observer using Events instead of callbacks.
```javascript
interface ThingObserver: EventTarget {
  void observe();  
    // observe options could be added for sampling/reporting frequency etc. 
  void unobserve();  // abort ongoing observation
  attribute EventHandler onchange;
  attribute EventHandler onerror;
  readonly attribute boolean observing; 
};

interface ObserveEvent: Event {
  readonly attribute any data;
};

ConsumedThing includes Observer;
ThingProperty includes Observer;
ThingEvent includes Observer;

partial interface WOT {
  ThingObserver discover(optional DiscoveryOptions options);
};
```
This one has the drawback to include too many things on ConsumedThing which may increase the likelihood to clash with property names defined by the [Thing Description](https://w3c.github.io/wot-thing-description/) that is used as a template when producing the ConsumedThing.

Nevertheless, for instance @danielpeintner prefers this solution for its simplicity to use.

**Note**: for both alternatives, because some implementations are using resource-constrained web runtimes, we'd need to use a subset of DOM Events that does not support capture/bubbling/cancelation, just `type` and `timestamp`, hence the `EventTarget` methods would also get simplified.

2. A syntactic variant using Events on a new Subscription interface (more aligned with e.g. the [Generic Sensor](https://www.w3.org/TR/generic-sensor/#idl-index)).
```javascript
interface Subscription { // subscribe methods return this object
  attribute EventHandler? ondata;  // SubscriptionEvent
  attribute ErrorHandler? onerror; // ErrorEvent
  readonly attribute boolean observing;
  void start();
  void stop();
};
interface SubscriptionEvent: Event {
  readonly attribute any data;
};
interface ErrorEvent: Event {
  readonly attribute Error error;
};
partial interface WoT {
  Subscription discover(optional ThingFilter filter); // .ondata.data is of type ThingDescription
};
partial interface ConsumedThing {
  Subscription observe(); //  ondata event .data is of type ThingDescription
};
partial interface ThingProperty {  
  Subscription observe(); //  ondata event .data: null; simple event; will update the hosting object
};
partial interface ThingEvent { 
  Subscription observe(); //  ondata event .data can be of any value, defined by the TD
};
```
This variant only adds a single method to e.g. ConsumedThing. Also, it is more aligned with Web NFC and Generic Sensor. However, it is more awkward to use it: call `observe()`, get a `Subscription`, set up event listeners, then call `start()` on the `Subscription`. Isn't then simpler to just pass callbacks (for handling data and errors) to `observe()`, like in next alternative?

3. Subscription with callbacks
```javascript
interface Subscription {
  Promise<void> start(NotificationHandler ondata);  // resolves when complete, rejects on irrecoverable error or when stopped
  void stop();
  readonly attribute boolean observing; 
};

callback NotificationHandler = void (any value);
callback ErrorHandler = void (Error error);

partial interface ConsumedThing {  // observed for TD change
  Subscription observe(); // ondata callback will get ThingDescription
};
partial interface ThingProperty {  // observed for value/state change
  Subscription observe(); // ondata will get no arguments
};
partial interface ThingEvent {  // observed for event data
  Subscription observe();  // ondata will get any value (defined by the TD)
};
partial interface WOT {
  Subscription discover(optional DiscoveryOptions options); // ondata will get ThingDescription
};
```

**Comment by zolkis** [comment]
@domenic @kenchris @cynthia could you please help us choosing how to go forward with this? 
Thank you!

**Comment by zolkis** [comment]
Thanks @domenic for the fast feedback. This API is shaky indeed for a number of reasons:
- there is a semantic overlap of WoT interactions (Thing Property, Thing Event, Thing Action) with script/object property, event, function
- we don't use Web IDL in the implementations (such as node-wot); we could instead use a TypeScript definition file; actually Web IDL brings a lot of complications.
- it tries to define an API to instantiate/represent objects defined by a JSON or JSON-LD Thing Description (TD), and add a few functions to model interaction methods defined by the TD. IOW it's very dynamic, property members are going to be defined in the future. Generic Sensor has picked a framework and defined a few sensor specs for defining the exact properties instead of an over-generalized approach found here. That rhymes with your suggestion to "step back" and deconstruct the API for individual use cases - which defies the purpose of the API.

For that reason the first question is whether do we really want to standardize such an API in W3C. The WG main deliverable is the [Thing Description spec](https://w3c.github.io/wot-thing-description/).

We could alternatively standardize not this API, but a higher level one that doesn't follow the Thing Description word by word.

**Comment by cynthia** [comment]
I'm missing quite a bit of context here; is this expected to run on a browsing context, or is this is an API for headless environment (e.g. node.js)? The latter probably doesn't quite fit in well with WebIDL semantics, although I have been told by people in TC39 and the Node community that they have been looking into using IDL - whether that will actually happen or is just a discussion is something I don't know.

I tend to agree with Domenic that some of these look like they'd make more sense to be events. The one question is the expected frequency of events, that would more or less be an important factor in deciding what would work best; a firehose of events is probably not something you want; at which point you should also wonder how to pull the processing off the main thread.

I'm not sure if I fully understand what `wot.discover()` is supposed to offer; the API naming suggests it would be a device discovery mechanism, and the choice of an observable feels a bit weird. I recall one example that was recently pushed to the platform, which does something similar, but not quite sure if this fits your case best: https://www.w3.org/TR/remote-playback/#dom-remoteplayback-watchavailability

(The promise+callback design feels rather odd, now that I revisit this though...)

**Comment by zolkis** [comment]
We tried to capture the common things in these interactions:
- start observing (a property change, a Thing Description change, a Thing Event) or start a discovery (which may work based on multicast request, or by checking a registry resource hosting a maintained list of discoverable items in the network)
- provide a callback to handle a change notification
- in the case of discovery, handle error notifications and completion notification (when supported)
- stop observing.

We could reverse this logic and start with separate solutions for each use case as Domenic suggested - maybe it's ok that they will look very similar, the spec will just get longer.

Indeed it wasn't fortunate to combine Promise with callback for discovery: a straight design with dedicated events (onfound, onerror, oncomplete) + start and stop methods would be clearer.

**Comment by zolkis** [comment]
I made a gist to experiment with an alternative API as per @domenic 's and @cynthia 's comments.
https://gist.github.com/zolkis/555441aebe1874392a00be50a8228f91 

On the other hand, we'll discuss in the group whether should we focus on this deliverable at all or leave it as a Node project ([node-wot](https://github.com/eclipse/thingweb.node-wot)) - I don't see much interest on a browser implementation (using HTTPS and WSS bindings for WoT). Indeed there have been opinions that the existing browser APIs are enough to handle WoT REST-ful interactions at lower level than this API (craft requests and process responses).

**Comment by zolkis** [comment]
Fixed in #165
